apollo-router/src/query_planner/fetch.rs

```diff
@@ -256,32 +451,95 @@ impl Variables {
-                    .get_key_value(key.as_str())
-                    .map(|(variable_key, value)| (variable_key.clone(), value.clone()))
-            }));
+        variables.extend(variable_usages.iter().filter_map(|key| {
+            body.variables
+                .get_key_value(key.as_str())
+                .map(|(variable_key, value)| (variable_key.clone(), value.clone()))
+        }));
 
+        if !requires.is_empty() {
             let mut inverted_paths: Vec<Vec<Path>> = Vec::new();
             let mut values: IndexSet<Value> = IndexSet::new();
-
+            let mut named_args: Vec<HashMap<String, Value>> = Vec::new();
             data.select_values_and_paths(schema, current_dir, |path, value| {
+                // first get contextual values that are required
+                let mut found_rewrites: HashSet<String> = HashSet::new();
+                let hash_map: HashMap<String, Value> = context_rewrites
+                    .iter()
+                    .flatten()
+                    .filter_map(|rewrite| {
+                        // note that it's possible that we could have multiple rewrites for the same variable. If that's the case, don't lookup if a value has already been found
+                        match rewrite {
+                            DataRewrite::KeyRenamer(item) => {
+                                if !found_rewrites.contains(item.rename_key_to.as_str()) {
+                                    let data_path = merge_context_path(path, &item.path);
+                                    let val = data_at_path(data, &data_path);
+                                    if let Some(v) = val {
+                                        // add to found
+                                        found_rewrites
+                                            .insert(item.rename_key_to.clone().to_string());
+                                        // TODO: not great
+                                        let mut new_value = v.clone();
+                                        if let Some(values) = new_value.as_array_mut() {
+                                            for v in values {
+                                                rewrites::apply_single_rewrite(
+                                                    schema,
+                                                    v,
+                                                    &DataRewrite::KeyRenamer({
+                                                        DataKeyRenamer {
+                                                            path: data_path.clone(),
+                                                            rename_key_to: item
+                                                                .rename_key_to
+                                                                .clone(),
+                                                        }
+                                                    }),
+                                                );
```

Comment:
forgot to comment on this before, but this new function seems to be doing what an existing function does on the `DataRewrite` impl - So I'd create a the `DataRewrite` first, and then run `maybe_apply` (will have to be made `pub(crate)`) on it:

```suggestion
                                                let data_rewrite = DataRewrite::KeyRenamer({
                                                    DataKeyRenamer {
                                                        path: data_path.clone(),
                                                        rename_key_to: item.rename_key_to.clone(),
                                                    }
                                                });
                                                data_rewrite.maybe_apply(schema, v);
```